libecole/include/ecole/environment/environment.hpp

```diff
@@ -112,22 +112,30 @@ class Environment {
 			dynamics().set_dynamics_random_state(model(), random_engine());
 
 			// Reset data extraction function and bring model to initial state.
-			observation_function().before_reset(model());
 			reward_function().before_reset(model());
+			observation_function().before_reset(model());
 			information_function().before_reset(model());
-			auto const [done, action_set] = dynamics().reset_dynamics(model(), std::forward<Args>(args)...);
+
+			// Place the environment in its initial state
+			auto&& [done, action_set] = dynamics().reset_dynamics(model(), std::forward<Args>(args)...);
 			can_transition = !done;
 
-			auto observation = OptionalObservation{};
-			if (!done) {
-				observation = observation_function().extract(model(), done);
-			}
+			// Extract reward, observation and information (in that order)
+			auto&& reward = reward_function().extract(model(), done);
+			auto&& observation = [&, done = done]() -> OptionalObservation {
```

Comment:
It's fine I guess, but isn't it a bit overkill to use an anonymous function here? Why not just a ternary operator `= ? :`

Reply:
Agreed, instantly invoked lambdas are only useful to initialize `const` variables. Also `done` is redudant with `&`.

Reply:
I agree it is a bit of an overkill, but I tried both and nonetheless I find this more readable. With ` = ? : ` you have to specify the class `OptionalObservation` twice, plus our code formatter does a terrible job at splitting the code on two lines. I let you judge:
```diff
-			auto&& observation = [&, done = done]() -> OptionalObservation {
-                               if (done) {  // Don't extract observations in final states
-                                       return {};
-                               }
-                               return observation_function().extract(model(), done);
-                       }();
+                       auto&& observation = done ? OptionalObservation{} :  // Don't extract observations in final states
+                                                                                                                OptionalObservation{observation_function().extract(model(), done)};
```